### PR TITLE
Added `gitlab_release_link` as a resource and data source

### PR DESCRIPTION
**Pull request by Moon1706**
## Description
Add resource for creation and management Gitlab release links (https://github.com/gitlabhq/terraform-provider-gitlab/issues/1035).

### PR Checklist Acknowledgement

- [x] I acknowledge that all of the following items are true, where applicable:
  - Resource attributes match 1:1 the names and structure of the API resource in [the GitLab API documentation](https://docs.gitlab.com/ee/api/).
  - [Examples](https://github.com/gitlabhq/terraform-provider-gitlab/tree/main/examples) are updated with:
    - A \*.tf file for the resource/s with at least one usage example
    - A \*.sh file for the resource/s with an import example (if applicable)
  - New resources have at minimum a basic test with three steps:
    - Create the resource
    - Update the attributes
    - Import the resource
  - No new `//lintignore` comments were copied from existing code. (Linter rules are meant to be enforced on new code.)

### CODE REVIEW

**Comment by Moon1706** [comment]
@timofurrer 
Review please. And check pr-acceptance-ee job.

File: internal/provider/data_source_gitlab_release_links.go
Comment (PatrickRice-KSC) [inline internal/provider/data_source_gitlab_release_links.go:61]: More a question for @timofurrer - Right now most of our "plural" data sources result in a 404 if they return 0 results due to how the underlying APIs work.. This can cause an issue in configurations where it's perfectly legit to have 0 results for a value. 

Is this really how we want these data sources to work, or should we catch 404 here using the `is404` function and return and break the loop which will just result in an empty list. The empty list will still cause an error in terraform configuration that _expect_ a value (because the downstream resources would break), but would still function perfectly fine in other scenarios where you may be just using the values as a loop in a `count` or `foreach` meta configuration.

Most of the other major cloud provider configurations are... inconsistent in this regard as well, with some data sources erroring, and some returning an empty array.

File: internal/provider/schema_gitlab_release_link.go
Comment (PatrickRice-KSC) [inline internal/provider/schema_gitlab_release_link.go:38]: Slight grammar correction 🙂 
```suggestion
			Description: "Relative path for a [Direct Asset link](https://docs.gitlab.com/ee/user/project/releases/index.html#permanent-links-to-release-assets).",
```

File: internal/provider/schema_gitlab_release_link.go
Comment (PatrickRice-KSC) [inline internal/provider/schema_gitlab_release_link.go:43]: Update slightly to match the verbiage we use for other similar lists 🙂 
```suggestion
			Description:      fmt.Sprintf("The type of the link. Valid values are %s. Defaults to %s.", renderValueListForDocs(validLinkTypes), validLinkTypes[0]),
```

**Review by PatrickRice-KSC — CHANGES_REQUESTED** [review]
@Moon1706  overall looking really good! I've requested a couple smaller changes, 1 to match a PR that was just merged, and 2 just around grammar and styling 🙂 . I left one comment for Timo that won't block merging just around general practices for data sources that we'll just resolve when it's the last comment left.

File: internal/provider/schema_gitlab_release_link.go
Comment (Moon1706) [reply internal/provider/schema_gitlab_release_link.go:38]: Okay)

File: internal/provider/data_source_gitlab_release_link_test.go
Comment (Moon1706) [reply internal/provider/data_source_gitlab_release_link_test.go:18]: Yea, it works with no problems.

File: internal/provider/data_source_gitlab_release_links.go
Comment (Moon1706) [reply internal/provider/data_source_gitlab_release_links.go:61]: Done. Additionally, I've added a test for this case.

**Comment by PatrickRice-KSC** [comment]
Restarted the test, since the failed test likely is a race condition due to container registry enablement being an async operation. We'll keep an eye on that if we see it elsewhere in PRs.

**Comment by PatrickRice-KSC** [comment]
Looks like we're golden. I'll give @timofurrer  a couple days to review (he tends to look for different things than I do 🙂 ) then we can merge.

**Review by PatrickRice-KSC — APPROVED** [review]

File: internal/provider/data_source_gitlab_release_links.go
Comment (timofurrer) [inline internal/provider/data_source_gitlab_release_links.go:22]: While the API "only" accepts URL-encoded paths, that's not the case for this provider, because `go-gitlab` will correctly URL-encode it for you. Let's use this instead:

```suggestion
				Description: "The ID or full path to the project.",
```

File: internal/provider/data_source_gitlab_release_links.go
Comment (timofurrer) [inline internal/provider/data_source_gitlab_release_links.go:75]: NIT:

```suggestion
	if err := d.Set("release_links", flattenGitlabReleaseLinks(project, tagName, releaseLinks)); err != nil {
		return diag.Errorf("Failed to set release links to state: %v", err)
	}
```

File: internal/provider/resource_gitlab_release_link.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_release_link.go:12]: NIT:

```suggestion
	"github.com/xanzy/go-gitlab"
```

File: internal/provider/resource_gitlab_release_link.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_release_link.go:17]:
```suggestion
		Description: `The ` + "`gitlab_release_link`" + ` resource allows to manage the lifecycle of a release link.
```

File: internal/provider/resource_gitlab_release_link.go
Comment (timofurrer) [inline internal/provider/resource_gitlab_release_link.go:41]: NIT:

```suggestion
		Name: gitlab.String(name),
		URL:  gitlab.String(url),
```

File: internal/provider/data_source_gitlab_release_links.go
Comment (timofurrer) [reply internal/provider/data_source_gitlab_release_links.go]: @PatrickRice-KSC @Moon1706 Yes, I agree to your judgement. My only concern where a `404` indeed is an error is if it is returned from any other page than the first, because that should actually be "accessible" - w.r.t. permissions and existence. 

But for the sake of simplicity I think it's "okay" to handle it as you've implemented it right now.

**Review by timofurrer — CHANGES_REQUESTED** [review]
@Moon1706 thanks for the nice work once again 💚 

I've left some feedback with suggestions to apply before a merge (require to re-generate docs, too) and some nitpicks. 

@PatrickRice-KSC thanks for the review! I've left an answer to the question you've asked :)

File: internal/provider/data_source_gitlab_release_links.go
Comment (Moon1706) [reply internal/provider/data_source_gitlab_release_links.go:22]: Done

File: internal/provider/data_source_gitlab_release_links.go
Comment (Moon1706) [reply internal/provider/data_source_gitlab_release_links.go]: @timofurrer If you want, I can append an additional section of condition (`if is404(err) && (options.Page == 1)`)

Comment (timofurrer) [reply internal/provider/data_source_gitlab_release_links.go]: @Moon1706 Yes, let's do that!

**Review by timofurrer — APPROVED** [review]
@Moon1706 It's an approve from my side once the additional condition from https://github.com/gitlabhq/terraform-provider-gitlab/pull/1131#discussion_r906648382 has been implemented 🎉 

Good work 🤝

File: internal/provider/data_source_gitlab_release_links.go
Comment (PatrickRice-KSC) [reply internal/provider/data_source_gitlab_release_links.go]: Agreed, I like this pattern!

Comment (Moon1706) [reply internal/provider/data_source_gitlab_release_links.go]: Done

**Review by timofurrer — APPROVED** [review]
Thanks @Moon1706 for the change and following through the feedback! 

Good work 🎉 🤝 

LGTM 🚢

**Pull request merged**
